src/prolly/map.ts

```diff
@@ -1,104 +1,131 @@
+  private _changedKeys: Set<string> = new Set();
+
+  constructor(items: Record<string, JSONValue> = {}) {
+    this._entries = Object.entries(items)
+        .map(e => deepFreeze(e) as Entry)
+        .sort((a, b) =>
+            stringCompare(a[0], b[0]),
+    );
   }
 
   static async load(hash: string, read: dag.Read): Promise<ProllyMap> {
     const chunk = await read.getChunk(hash);
     if (chunk === undefined) {
       throw new Error(`Chunk not found: ${hash}`);
     }
-    const base = Leaf.load(chunk);
-    return new ProllyMap(base, new Map());
-  }
 
-  has(key: string): boolean {
-    const p = this._pending.get(key);
-    if (p !== undefined) {
-      return p !== deleteSentinel;
+    // Validate at load-time so we can assume data is valid thereafter.
+    const entries = chunk.data;
+    // Assert that the shape/type is correct
+    assertEntries(entries);
+    // But also assert that entries is sorted and has no duplicate keys.
+    const seen = new Set();
+    for (let i = 0; i < entries.length - 1; i++) {
+      const entry = entries[i];
+      const next = entries[i + 1];
+      if (entry[0] === next[0] || seen.has(entry[0])) {
+        throw new Error('duplicate key');
+      }
+      if (entry[0] > next[0]) {
+        throw new Error('unsorted key');
+      }
+      seen.add(entry[0]);
     }
 
-    return this._baseHas(key);
+    const map = new ProllyMap();
+    map._entries = entries;
+    return map;
   }
 
-  private _baseHas(key: string): boolean {
-    if (this._base === undefined) {
-      return false;
+  private _binarySearch(key: string): {found: boolean; index: number} {
+    const entries = this._entries;
+    let size = entries.length;
+    if (size === 0) {
+      return {found: false, index: 0};
+    }
+    let base = 0;
+
+    while (size > 1) {
+      const half = Math.floor(size / 2);
+      const mid = base + half;
+      // mid is always in [0, size), that means mid is >= 0 and < size.
+      // mid >= 0: by definition
+      // mid < size: mid = size / 2 + size / 4 + size / 8 ...
+      const entry = entries[mid];
+      // No way that key can be None.
+      const cmp = stringCompare(entry[0], key);
+      base = cmp > 0 ? base : mid;
+      size -= half;
+    }
+    // base is always in [0, size) because base <= mid.
+    const entry = entries[base];
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const cmp = stringCompare(entry[0], key);
+    if (cmp === 0) {
+      return {found: true, index: base};
     }
-    return this._base.binarySearch(key).found;
+    return {found: false, index: base + cmp};
   }
 
-  get(key: string): JSONValue | undefined {
-    const ks = key;
-    const p = this._pending.get(ks);
-    switch (p) {
-      case deleteSentinel:
-        return undefined;
-      case undefined:
-        return this._baseGet(key);
-      default:
-        return p;
-    }
+  has(key: string): boolean {
+    const {found} = this._binarySearch(key);
```

Comment:
inline

Reply:
This isn't in the newer PR, but for my info, can you explain what you mean here? I don't understand your comment.

Reply:
I just wanted it to be: `return this._binarySearch(key).found`

BTW. I'm applying this change.